src/v/cluster/partition_leaders_table.h

```diff
@@ -193,27 +171,87 @@ class partition_leaders_table {
         model::revision_id partition_revision;
     };
 
-    std::optional<leader_meta>
+    using partition_leaders = absl::btree_map<model::partition_id, leader_meta>;
+    using topics_t = absl::node_hash_map<
+      model::topic_namespace,
+      partition_leaders,
+      model::topic_namespace_hash,
+      model::topic_namespace_eq>;
+
+    std::optional<std::reference_wrapper<const leader_meta>>
       find_leader_meta(model::topic_namespace_view, model::partition_id) const;
 
-    absl::node_hash_map<leader_key, leader_meta, leader_key_hash, leader_key_eq>
-      _leaders;
+    void throw_if_modified(version current_version) const {
+        if (unlikely(current_version != _version)) {
+            throw concurrent_modification_error(current_version, _version);
+        }
+    }
+
+    void do_update_partition_leader(
+      bool is_controller,
+      topics_t::iterator,
+      model::partition_id,
+      model::revision_id,
+      model::term_id,
+      std::optional<model::node_id>);
 
-    // per-ntp notifications for leadership election. note that the
-    // namespace is currently ignored pending an update to the metadata
-    // cache that attaches a namespace to all topics partition references.
-    int32_t _promise_id = 0;
-    using promises_t = absl::node_hash_map<
-      model::ntp,
-      absl::node_hash_map<
-        int32_t,
-        std::unique_ptr<expiring_promise<model::node_id>>>>;
+    topics_t _topic_leaders;
 
-    promises_t _leader_promises;
+    uint64_t _leaderless_partition_count{0};
 
     ss::sharded<topic_table>& _topic_table;
 
     ntp_callbacks<leader_change_cb_t> _watchers;
+    /**
+     * Store version to check for concurrent updates, when version was
+     * incremented while iterating over the list of leaders
+     */
+    version _version{0};
+    version _topic_map_version{0};
+    ss::gate _gate;
+    ss::abort_source _as;
 };
+namespace internal {
+/**
+ * Yields every n calls of maybe_yield. Used to prevent calling
+ * `ss::coroutine::maybe_yield` as it is expensive.
+ */
+class yield_tracker {
+public:
+    explicit yield_tracker(size_t check_every)
+      : _check_every(check_every) {}
+
+    ss::future<> maybe_yield() {
```

Comment:
I am not sure this pattern actually works as is as the expensive bit is really the coroutine machinery so now we would just call this `maybe_yield` which would suffer from the same issue as if we called `ss::coroutine::maybe_yield`. 

Ideally for a generic solution we probably want something like:

```
template<typename Iter, typename Func>
ss::future<> yield_every(Iter begin, Iter end, size_t every, Func func) {
    while (true) {
        for (size_t i = 0; i < every; ++i, ++begin) {
            if (begin == end) {
                co_return;
            }
            func(*begin);
        }

        co_await ss::coroutine::maybe_yield();
    }
}
```

But this obviously gets tricky if you have nested loops.

Reply:
Yes exactly. The key is that in the tight loop (which executes once per element) there should be not be any co_await (or equivalent continuation-based thing), just plan iterator stuff. Then you break out periodically do to the maybe yield like in @StephanDollberg example.

It's not only the _runtime_ call to `maybe_yield` that is a bit expensive but just having in the tight loop, even if it is skipped most iteration, generates a lot of suspend boilerplate that harms optimization.